src/main/java/uk/ac/ic/wlgitbridge/bridge/db/postgres/PostgresDBStore.java

```diff
@@ -0,0 +1,298 @@
+package uk.ac.ic.wlgitbridge.bridge.db.postgres;
+
+import uk.ac.ic.wlgitbridge.bridge.db.DBInitException;
+import uk.ac.ic.wlgitbridge.bridge.db.DBStore;
+import uk.ac.ic.wlgitbridge.bridge.db.ProjectState;
+import uk.ac.ic.wlgitbridge.util.Log;
+
+import java.sql.*;
+import java.util.ArrayList;
+import java.util.List;
+
+public class PostgresDBStore implements DBStore {
+
+  private final Connection connection;
+
+  public PostgresDBStore(String connectionString, String user, String pass) {
+    Log.info("Initialize PostgresDBStore");
+    try {
+      Class.forName("org.postgresql.Driver");
+      connection = DriverManager
+        .getConnection(connectionString, user, pass);
+    } catch (Exception e) {
+      Log.error("Error connecting to Postgres: {}", e.getMessage());
+      throw new DBInitException(e);
+    }
+  }
+
+  public void dangerousResetDatabaseDontDoIt() {
+    try {
+      Statement statement = connection.createStatement();
+      statement.execute("delete from url_index_store;");
+      statement.execute("delete from projects;");
+      statement.close();
+    } catch (Exception e) {
+      Log.error("Error while querying Postgres: {}", e.getMessage());
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public int getNumProjects() {
+    try {
+      Statement statement = connection.createStatement();
+      ResultSet rs = statement.executeQuery("SELECT count(*) from projects");
+      if (rs.next()) {
+        int result = rs.getInt(1);
+        rs.close();
+        statement.close();
+        return result;
+      } else {
+        throw new Exception("No results from query");
```

Comment:
Would we still need to `statement.close()` (and `rs.close()` maybe) here?

I can't see why `count(*)` would fail to return a value, but given that we are these up above before returning, I'm wondering whether we need to do that here as well.

Reply:
Yup, I think we should put these in `finally` blocks. :+1:

Reply:
Vexingly, `.close()` can throw an exception, and so when it's placed in a `finally` block the compiler wants us to add the exception signature to the method, or to wrap it in yet another `try-catch` block...

Reply:
try-with-resources seems to work well :+1: